### DISCUSSION TITLE
C# unions?

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#1554] [Language Ideas]
This probably isn't something I'd ordinarily suggest for a language feature but it is possible to implement it now we have C# 7.3

Do others think there is any sense in seeking to make this more of an inherent C# feature though:

```
namespace NewConsole
{
    unsafe class Program
    {
        static void Main(string[] args)
        {
            Union union = new Union(null); // pass a real address here in practice!

            // layout1 and layout2 each refer to the same memory address.

            var layout1 = union.AccessAs<Layout_1>();
            var layout2 = union.AccessAs<Layout_2>();

            layout1.Counter = 23;
            layout2.Rate = 34.5;
        }
    }
    public unsafe class Union
    {
        public void* Address { get; private set; }
        public Union (void * Address)
        {
            this.Address = Address;
        }

        public ref T AccessAs<T>() where T : unmanaged
        {
            return ref (*(T*)Address);
        }
    }
    public struct Layout_1
    {
        public long Counter;
        public int Depth;
        public int Other;
    }
    public struct Layout_2
    {
        public byte Byte1;
        public byte Byte2;
        public byte Byte3;
        public byte Byte4;
        public double Rate;
    }
}
```

The alignmnet of each struct must of course be the same and a few other considerations...

**Comment by tannergooding** [comment]
See also: https://github.com/dotnet/csharplang/issues/687

As a workaround, you can already declare unions today using `[StructLayout(LayoutKind.Explicit)]` and `[FieldOffset(int)]`

**Comment by tannergooding** [comment]
```C#
using System.Runtime.InteropServices;

[StructLayout(LayoutKind.Explicit)]
public struct Union
{
    [FieldOffset(0)]
    public Layout_1 Layout_1;

    [FieldOffset(0)]
    public Layout_2 Layout_2;
}

public struct Layout_1
{
    public long Counter;
    public int Depth;
    public int Other;
}

public struct Layout_2
{
    public byte Byte1;
    public byte Byte2;
    public byte Byte3;
    public byte Byte4;
    public double Rate;
}
```

**Comment by tannergooding** [comment]
See also https://github.com/dotnet/csharplang/issues/399

**Comment by tannergooding** [comment]
and https://github.com/dotnet/csharplang/issues/113

**Comment by The-Futurist** [comment]
@tannergooding - Yep, you're right - I've never had ocassion to use a "union" in a .Net world (but there was a question about it on MSDN Forums earlier). The attributes are clearly a great way to go, however the generic Union class/method is _generic_ and there's no intermediate union stuct declaring specific members.

**Comment by The-Futurist** [comment]
This includes some alignment capabilities:

```
using System;

namespace NewConsole
{
    unsafe class Program
    {
        static void Main(string[] args)
        {
            Union union = new Union(null, 16);

            // layout1 and layout2 each refer to the same memory address.

            var layout1 = union.AccessAs<Layout_1>();
            var layout2 = union.AccessAs<Layout_2>();

            layout1.Counter = 23;
            layout2.Rate = 34.5;
        }
    }
    public unsafe class Union
    {
        public void* Address { get; private set; }
        public int Alignment { get; private set; }
        public Union (void * Address, int Alignment)
        {
            if (Address == null)
                throw new ArgumentNullException("The supplied address must not be null.");

            // sanity check Alignment...

            if (Alignment != 1 && Alignment != 2 && Alignment != 4 && Alignment != 8 && Alignment != 16)
                throw new ArgumentException("The supplied alignment may only be 1, 2, 4, 8, or 16.");

            // sanity check the Address...

            if ((long)(Address) % Alignment != 0)
                throw new ArgumentException("The supplied address is not aligned as expected.");

            this.Address = Address;
            this.Alignment = Alignment;
        }

        public ref T AccessAs<T>() where T : unmanaged
        {
            // check that alignment of T matches the Alignment value...

            return ref (*(T*)Address);
        }
    }
    public struct Layout_1
    {
        public long Counter;
        public int Depth;
        public int Other;
    }
    public struct Layout_2
    {
        public byte Byte1;
        public byte Byte2;
        public byte Byte3;
        public byte Byte4;
        public double Rate;
    }
}
```

**Comment by The-Futurist** [comment]
Actually this is what the main method _should_ look like:

```
        static void Main(string[] args)
        {
            Union union = new Union(null, 16);

            // layout1 and layout2 each refer to the same memory address.

            ref var layout1 = ref union.AccessAs<Layout_1>();
            ref var layout2 = ref union.AccessAs<Layout_2>();

            layout1.Counter = 23;
            layout2.Rate = 34.5;
        }

```
I should have got that right since I posted this https://github.com/dotnet/csharplang/issues/1553

**Comment by huoyaoyuan** [comment]
Why don't you use common base class and pattern matching? For performence critical code, unsafe should be the right way.

**Comment by tannergooding** [comment]
@Korporal, what is the purpose of the wrapper `Union` class? Why not just do:
```C#
using System;
using System.Runtime.InteropServices;

class Program
{
    static void Main(string[] args)
    {
        Variant1();
        Console.WriteLine();
        Variant2();
    }

    static void Variant1()
    {
        var union = new Union();

        Console.WriteLine(union.layout1);   // Counter: 0; Depth: 0; Other: 0;
        Console.WriteLine(union.layout2);   // Byte1: 0; Byte2: 0; Byte3: 0; Byte4: 0; Rate: 0;

        union.layout1.Counter = 23;
        union.layout2.Rate = 34.5;

        Console.WriteLine(union.layout1);   // Counter: 23; Depth: 0; Other: 1078018048;
        Console.WriteLine(union.layout2);   // Byte1: 23; Byte2: 0; Byte3: 0; Byte4: 0; Rate: 34.5;
    }

    static void Variant2()
    {
        var union = new Union();

        ref var layout1 = ref union.layout1;
        ref var layout2 = ref union.layout2;

        Console.WriteLine(union.layout1);   // Counter: 0; Depth: 0; Other: 0;
        Console.WriteLine(union.layout2);   // Byte1: 0; Byte2: 0; Byte3: 0; Byte4: 0; Rate: 0;

        layout1.Counter = 23;
        layout2.Rate = 34.5;

        Console.WriteLine(union.layout1);   // Counter: 23; Depth: 0; Other: 1078018048;
        Console.WriteLine(union.layout2);   // Byte1: 23; Byte2: 0; Byte3: 0; Byte4: 0; Rate: 34.5;

    }
}

[StructLayout(LayoutKind.Explicit)]
public struct Union
{
    [FieldOffset(0)]
    public Layout_1 layout1;

    [FieldOffset(0)]
    public Layout_2 layout2;
}

public struct Layout_1
{
    public long Counter;
    public int Depth;
    public int Other;

    public override string ToString()
    {
        return $"Counter: {Counter}; Depth: {Depth}; Other: {Other};";
    }
}

public struct Layout_2
{
    public byte Byte1;
    public byte Byte2;
    public byte Byte3;
    public byte Byte4;
    public double Rate;

    public override string ToString()
    {
        return $"Byte1: {Byte1}; Byte2: {Byte2}; Byte3: {Byte3}; Byte4: {Byte4}; Rate: {Rate};";
    }
}
```

**Comment by The-Futurist** [comment]
@tannergooding - Nothing wrong with that code BUT the "Union" struct in this case is inherently specific to the two data structs. The "Union" wrapper class in my example is truly generic and requires no structs other that the actual data structs (Layout_1 and Layout_2) themselves.

**Comment by tannergooding** [comment]
> The "Union" wrapper class in my example is truly generic and requires no structs other that the actual data structs (Layout_1 and Layout_2) themselves.

Just use `Unsafe.As`?

```C#
using System;
using System.Runtime.CompilerServices;

class Program
{
    static void Main(string[] args)
    {
        var layout1 = new Layout_1();
        ref var layout2 = ref Unsafe.As<Layout_1, Layout_2>(ref layout1);

        Console.WriteLine(layout1);   // Counter: 0; Depth: 0; Other: 0;
        Console.WriteLine(layout2);   // Byte1: 0; Byte2: 0; Byte3: 0; Byte4: 0; Rate: 0;

        layout1.Counter = 23;
        layout2.Rate = 34.5;

        Console.WriteLine(layout1);   // Counter: 23; Depth: 0; Other: 1078018048;
        Console.WriteLine(layout2);   // Byte1: 23; Byte2: 0; Byte3: 0; Byte4: 0; Rate: 34.5;
    }
}

public struct Layout_1
{
    public long Counter;
    public int Depth;
    public int Other;

    public override string ToString()
    {
        return $"Counter: {Counter}; Depth: {Depth}; Other: {Other};";
    }
}

public struct Layout_2
{
    public byte Byte1;
    public byte Byte2;
    public byte Byte3;
    public byte Byte4;
    public double Rate;

    public override string ToString()
    {
        return $"Byte1: {Byte1}; Byte2: {Byte2}; Byte3: {Byte3}; Byte4: {Byte4}; Rate: {Rate};";
    }
}
```

**Comment by tannergooding** [comment]
`Unsafe.AsRef` also has another overload which takes a void* and returns a `ref T`, so you would do `ref var layout2 = ref Unsafe.AsRef<Layout_2>(&layout1)`

**Comment by The-Futurist** [comment]
@tannergooding - If you're referring to [this code](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il) (which I'd never seen before and seems to be written in IL) then you may be correct. But it does not do any alignmnet checks (but these could be added in principle). That also seems to be "lets refer to struct A as struct B" so is inherently limited to seeing one struct in terms of another rather than "lets refer to the addres as struct A" or "lets refer to the address as struct B" and so on. But otherwise it seems very similar indeed!

Incidentally I wrote a library that did these and several similar things some years ago (by dynamically generating code form IL) and did so mostly because of the lack of the recent "ref" and "unmanaged" support, so glad to see this!

**Comment by tannergooding** [comment]
> But it does not do any alignmnet checks (but these could be added in principle).

Yes, you can write your own wrapper to do the alignment checks. However, I'm not sure why alignment checks would be required. The generated IL should work, regardless of how the types are aligned (and in practice the JIT won't let you specify how you want types to be aligned anyways, you need a custom allocator for that).

> hat also seems to be "lets refer to struct A as struct B" so is inherently limited to seeing one struct in terms of another rather than "lets refer to the addres as struct A" or "lets refer to the address as struct B" and so on. But otherwise it seems very similar indeed!

A ref is just a type-safe (and GC-safe) way of saying this is the address of a given type. So the code is exactly saying "lets treat this reference (managed pointer) to struct A as a reference (managed pointer) to struct B".

**Comment by HaloFour** [comment]
Outside of specific interop/marshaling scenarios or for narrow performance use cases is there a lot of reason to try to map the multiple types within a union into the same memory space?

I'd rather that energy be put into type union/intersections and discriminated unions.  I think that those features would have significantly more reach, and neither should require trying to hack around with pointers and unsafe code.

**Comment by ufcpp** [comment]
See the last section of https://github.com/dotnet/csharplang/blob/master/meetings/2018/LDM-2018-05-14.md

> We could work with the runtime to allow us to overlay different types inside of a struct. For unions where each variant is small, structs would probably be highly preferable from a performance perspective.

It requires CLR changes.

**Comment by ufcpp** [comment]
see also https://github.com/dotnet/csharplang/issues/317

**Comment by tannergooding** [comment]
@ufcpp, it doesn't require CLR changes, it (and other languages already exposing support for discriminated unions) would just benefit from them.

**Comment by ufcpp** [comment]
@tannergooding 
Class unions (F#'s discriminated union) does not but struct unions requires it, especially generic struct unions.

```cs
using System;
using System.Runtime.InteropServices;

// OK
[StructLayout(LayoutKind.Explicit)]
readonly struct Union_Int_DateTime
{
    [FieldOffset(0)]
    public readonly int Discriminator;

    [FieldOffset(4)]
    public readonly int Item1;

    [FieldOffset(4)]
    public readonly DateTime Item2;

    public Union_Int_DateTime(int value) : this() => (Discriminator, Item1) = (0, value);
    public Union_Int_DateTime(DateTime value) : this() => (Discriminator, Item2) = (1, value);
}

// OK
[StructLayout(LayoutKind.Explicit)]
readonly struct Union_String_StringArray
{
    [FieldOffset(0)]
    public readonly int Discriminator;

    [FieldOffset(4)]
    public readonly string Item1;

    [FieldOffset(4)]
    public readonly string[] Item2;

    public Union_String_StringArray(string value) : this() => (Discriminator, Item1) = (0, value);
    public Union_String_StringArray(string[] value) : this() => (Discriminator, Item2) = (1, value);
}

// Compilable but runtime exception
[StructLayout(LayoutKind.Explicit)]
readonly struct Union_String_Int
{
    [FieldOffset(0)]
    public readonly int Discriminator;

    [FieldOffset(4)]
    public readonly string Item1;

    [FieldOffset(4)]
    public readonly int Item2;

    public Union_String_Int(string value) : this() => (Discriminator, Item1) = (0, value);
    public Union_String_Int(int value) : this() => (Discriminator, Item2) = (1, value);
}

// Compilable but runtime exception
[StructLayout(LayoutKind.Explicit)]
struct Union<T1, T2>
{
    [FieldOffset(0)]
    public readonly int Discriminator;

    [FieldOffset(4)]
    public readonly T1 Item1;

    [FieldOffset(4)]
    public readonly T2 Item2;

    public Union(T1 value) : this() => (Discriminator, Item1) = (0, value);
    public Union(T2 value) : this() => (Discriminator, Item2) = (1, value);
}
```

**Comment by The-Futurist** [comment]
@HaloFour - Hi, I guess this is pretty much moot anyway since Tanner's option is more or less fine for this and my approach is another alternative each of these is trivial to use in code, so I agree, with hindsight this is too esoteric for a candidate language feature.

**Comment by bbarry** [comment]
I am not sure if it is too esoteric. I think various members of `Unsafe` and new coming members of `MemoryMarshal` provide basically what we want (as far as what can be probably ok from the CLR) at an api level for structural access of allocated memory. I think there might be certain patterns around those apis that deserve to be codified into some syntax sugar in order to bless them as correct and intended (that is why we have `lock`, `using`, `async` and a few other syntax constructs after all).


It could be useful maybe to have something like


    var layout1 = new Layout_1();
    union(ref Layout_2 layout2 = ref layout1)
    {
        ...
    }


Some of these "new" low level types and methods (and the `ref` everywhere syntaxes) are (quite rightly) intimidating to users. If there are general patterns of reasonably useful things to do with them that can be done with a higher level syntax or library they should be considered for either language or the core libraries.
